### PR TITLE
 Fix in TSGForOI muons - 93X

### PR DESCRIPTION
**Pull request by silviodonato**
92X version: #19947
------------------------------------------
Running hltIntegrationTests /dev/CMSSW_9_2_0/GRun/V94 -i root://cms-xrd-global.cern.ch//store/data/Run2017B/MuonEG/RAW/v1/000/297/469/00000/C0653517-2D59-E711-91CD-02163E0118F3.root -x "--globaltag 92X_dataRun2_HLT_v4" -x "--l1-emulator uGT --l1 L1Menu_Collisions2017_v2_m6_full_xml" -j16 -n100 -d hltintegrationTestV94 &>logIntegrationDataV94 &
we found inconsistencies between running full HLT menu and single HLT path.

This is due to some non-const members of TSGForOI that are not initialized at each event.

This PR is very important for the upcoming V2.1 menu.

### CODE REVIEW

**Comment by silviodonato** [comment]
please test

**Comment by Martin-Grunewald** [comment]
type bug-fix

**Comment by Martin-Grunewald** [comment]
urgent

**Comment by slava77** [comment]
is v2.1 menu in the release already to have this PR actually tested with jenkins.
It seems from the PR description that the issue is related in part to the configuration of this module
[or configuration of the tests used previously that never enabled it]

**Comment by silviodonato** [comment]
Yes, the v2.1 HLT menu has been integrated in release with #19876.
Unfortunately, I think hltIntegrationTests in  jenkins cannot spot this issue because you need to more statistic to see it.

**Comment by slava77** [comment]
@cmsbuild please test

@smuzaffar I see  "Schedule cmssw#19948" in the scheduling jobs
https://cmssdt.cern.ch/jenkins/job/ib-schedule-pr-tests/19497/console
but nothing in ib-any-integration
Please check what happened.

**Comment by slava77** [comment]
.. OK, this time it actually worked 
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/21855/console

**Comment by silviodonato** [comment]
hltIntegrationTests are not run in jenkins ( https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19876/21703/addOnTests/ ), so they were not able to spot this issue.

**Comment by silviodonato** [comment]
hltIntegrationTests in IB validation: https://cmssdt.cern.ch/SDT/jenkins-artifacts/HLT-Validation/CMSSW_9_3_X_2017-07-27-2300/slc6_amd64_gcc630/

**Comment by Martin-Grunewald** [comment]
They are not run in PR/jenkins tests but in IB tests: "HLT Validation".
The statistics (-n 100) was still too small to detect that one, though.

**Comment by Martin-Grunewald** [comment]
The failed tests are due to the bug introduced with https://github.com/cms-sw/cmssw/pull/19824
and fixed by https://github.com/cms-sw/cmssw/pull/19951

**Comment by smuzaffar** [comment]
please test
this should now use CMSSW_9_3_X_2017-07-27-2300 for tests

**Comment by slava77** [comment]
assign hlt
partly to have appropriate label in the release notes
partly because proper checkout is better done/confirmed by an HLT L2 (even though Silvio is one of the L2s)

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 1]
```diff
@@ -52,13 +50,31 @@ TSGForOI::~TSGForOI(){
...
+  layerCount_ = 0;
+
+  /// Surface used to make a TSOS at the PCA to the beamline
+  Plane::PlanePointer dummyPlane_ = Plane::build(Plane::PositionType(), Plane::RotationType());
+
+  /// Read ESHandles
+  edm::ESHandle<MagneticField>          magfield_;
+  edm::ESHandle<Propagator>             propagatorAlong_;
+  edm::ESHandle<Propagator>             propagatorOpposite_;
+  edm::ESHandle<GlobalTrackingGeometry> geometry_;
```
Comment (slava77) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: please remove trailing underscore to be clear that these are local variables

**Comment by slava77** [comment]
it looks like TSGForOI is now about 10 lines of code away from becoming ::global
Any reason not to do it already?

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
@slava77 I move this module to ::global. It was not so trivial since I had to adapt class functions to avoid to class variables. I've checked that after this commit trigger results are unchanged.

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 2]
```diff
@@ -130,43 +133,43 @@ void TSGForOI::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
 
     //		BARREL
     if (std::abs(l2->eta()) < maxEtaForTOB_) {
-      layerCount_ = 0;
+      layerCount = 0;
       for (auto it=tob.rbegin(); it!=tob.rend(); ++it) {	//This goes from outermost to innermost layer
-	LogTrace("TSGForOI") << "TSGForOI::produce: looping in TOB layer " << layerCount_ << endl; 
-	findSeedsOnLayer(tTopo, **it, tsosAtIP,  *(propagatorAlong.get()), *(propagatorOpposite.get()), l2, out);
+	LogTrace("TSGForOI") << "TSGForOI::produce: looping in TOB layer " << layerCount << endl; 
+	findSeedsOnLayer(tTopo, **it, tsosAtIP,  *(propagatorAlong.get()), *(propagatorOpposite.get()), l2, estimatorH, measurementTrackerH, numSeedsMade, numOfMaxSeeds, layerCount, foundHitlessSeed, analysedL2, out);
```
Comment (slava77) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: a line break after ```l2,``` or so should improve readability

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 3]
```diff
@@ -130,43 +133,43 @@ void TSGForOI::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
 
     //		ENDCAP+
     if (l2->eta() > minEtaForTEC_) {
-      layerCount_ = 0;
+      layerCount = 0;
       for (auto it=tecPositive.rbegin(); it!=tecPositive.rend(); ++it) {
-	LogTrace("TSGForOI") << "TSGForOI::produce: looping in TEC+ layer " << layerCount_ << endl; 
-	findSeedsOnLayer(tTopo, **it, tsosAtIP,  *(propagatorAlong.get()), *(propagatorOpposite.get()), l2, out);
+	LogTrace("TSGForOI") << "TSGForOI::produce: looping in TEC+ layer " << layerCount << endl; 
+	findSeedsOnLayer(tTopo, **it, tsosAtIP,  *(propagatorAlong.get()), *(propagatorOpposite.get()), l2, estimatorH, measurementTrackerH, numSeedsMade, numOfMaxSeeds, layerCount, foundHitlessSeed, analysedL2, out);
```
Comment (slava77) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: please add a line break here and also below

**Comment by silviodonato** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

the last one failed with 
```Build timed out (after 360 minutes). Marking the build as aborted.```

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19948 644c0b8
- bug fixes in TSGForOI and migration of the module to a global 
- jenkins tests pass  and comparisons show no differences since e04a72c
    - local test with 136.782 (double muon 2017)  confirms that there were no changes introduced since e04a72c
- local tests with 2017 workflows show that the reco quantities are not affected and DQM comparisons with baseline show changes in HLT muon-related quantities expected from changes in this PR.

**Comment by Martin-Grunewald** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**